content/doc/book/system-administration/authenticating-scripted-clients.adoc

```diff
@@ -0,0 +1,378 @@
+----
+
+For additional information, see the
+https://github.com/cdancy/jenkins-rest/wiki[cdancy/jenkins-rest wiki].
+
+[[Authenticatingscriptedclients-Groovyscriptusingcommons-httpclient]]
+== Groovy script using commons-httpclient
+
+In a groovy script, this could look something like this (using
+commons-httpclient):
+
+[source,groovy]
+----
+import org.apache.commons.httpclient.*
+
+import org.apache.commons.httpclient.auth.*
+import org.apache.commons.httpclient.methods.*
+
+@Grab(group='commons-httpclient', module='commons-httpclient', version='3.1')
+void createNewJenkinsProject() {
+
+  def server = "server"
+  def jenkinsHost = "https://${server}/jenkins/"
+  def projectName = "TEST"
+  def configurationFile = "config.xml"
+
+  def username = "username"
+  def apiToken = "apiToken"
+
+  def client = new HttpClient()
+  client.state.setCredentials(
+    new AuthScope( server, 443, "realm"),
+    new UsernamePasswordCredentials( username, apiToken )
+  )
+
+  // Jenkins does not do any authentication negotiation,
+  // i.e. it does not return a 401 (Unauthorized)
+  // but immediately a 403 (Forbidden)
+  client.params.authenticationPreemptive = true
+
+  def post = new PostMethod( "${jenkinsHost}/createItem?name=${projectName}" )
+  post.doAuthentication = true
+
+  File input = new File(configurationFile);
+  RequestEntity entity = new FileRequestEntity(input, "text/xml; charset=UTF-8");
+  post.setRequestEntity(entity);
+  try {
+    int result = client.executeMethod(post)
+    println "Return code: ${result}"
+    post.responseHeaders.each{ println it.toString().trim() }
+    println post.getResponseBodyAsString()
+  } finally {
+    post.releaseConnection()
+  }
+}
+createNewJenkinsProject()
+----
+
+
+NOTE: If you are using wget 1.11 against Jenkins version 1.586 and above
+with the
+https://issues.jenkins-ci.org/browse/JENKINS-25169[JENKINS-25169] fix,
+you might need to use the following options:
+
+[source,sh]
+----
+wget --auth-no-challenge \
+    --http-user=user --http-password=apiToken \
+    --secure-protocol=TLSv1 \
+    http://jenkins.yourcompany.com/job/your_job/build?token=TOKEN
+----
+
+NOTE: If you are using wget 1.11, you might need to use the following
+options:
+
+[source,sh]
+----
+ wget --auth-no-challenge \
+     --http-user=user \
+     --http-password=apiToken \
+     http://jenkins.yourcompany.com/job/your_job/build?token=TOKEN
+----
+
+With wget 1.10.x the following is enough (but will not work with 1.11.x)
+:
+
+[source,sh]
+----
+ wget http://user:apiToken@jenkins.yourcompany.com/job/your_job/build?token=TOKEN
+----
+
+See this RedHat bug report for more detailled explanations:
+https://bugzilla.redhat.com/show_bug.cgi?id=446949 (this also affect
+other distributions) +
+(Report indicates that wget 1.0 and 1.1 don't appear to have a
+--auth-no-challenge option.)
+
+[[Authenticatingscriptedclients-PerlLWPexampleforascriptedclient]]
+== Perl LWP example for a scripted client
```

Comment:
is perl a useful example these days?

Reply:
I think we should leave it as a "retro example".  I suspect there are still plenty of places that use perl for powerful text processing.